### PR TITLE
[SPARK-34072][SQL] Fix empty array failed in functions.lit()

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add `java.lang.Object` type check in `componentTypeToDataType `.

### Why are the changes needed?

We will get exception if we execute code:
`spark.range(1).select(lit(Array()))`
error msg:
```
Exception in thread "main" org.apache.spark.sql.AnalysisException: Unsupported component type class java.lang.Object in arrays;
	at org.apache.spark.sql.catalyst.expressions.Literal$.componentTypeToDataType(literals.scala:129)
	at org.apache.spark.sql.catalyst.expressions.Literal$.apply(literals.scala:80)
	at org.apache.spark.sql.catalyst.expressions.Literal$.$anonfun$create$2(literals.scala:148)
	at scala.util.Failure.getOrElse(Try.scala:222)
	at org.apache.spark.sql.catalyst.expressions.Literal$.create(literals.scala:148)
	at org.apache.spark.sql.functions$.typedLit(functions.scala:132)
	at org.apache.spark.sql.functions$.lit(functions.scala:115)
```

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Add test.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 1]
```diff
@@ -1982,6 +1982,13 @@ class DatasetSuite extends QueryTest
       assert(timezone == "Asia/Shanghai")
     }
   }
+
+  test("SPARK-34072: Fix empty array failed in functions.lit()") {
+    checkAnswer(
+      spark.range(1).select(lit(Array())),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1988]: Can't you create an array such as: 

```scala
scala> lit(Array.empty[Int])
res0: org.apache.spark.sql.Column = []
```

? Matching `java.lang.Object` to `NullType` is counterintuative.

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1988]: The problem is `Literal.create(Array()).dataType` is also `ArrayType(NullType)`. `lit` just lost the `ClassTag`.

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1988]: BTW `HiveInspectors.javaTypeToDataType` also has this pattern `case c: Class[_] if c == classOf[java.lang.Object] => NullType`.

Comment (MrPowers) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1988]: @ulysses-you - does `spark.range(1).select(typedLit(Array()))` get you the result you're looking for?  Per the docs, "The difference between the typedLit function and lit is that this function can handle parameterized scala types e.g.: List, Seq and Map."  `lit` isn't normally used for collections.

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1988]: Yes the right way is using `typedLit`. I just consider the issue is easy to resolve so maybe we can support it directly.

**Pull request closed without merging**